### PR TITLE
Cuda support for SMPC in 0.2.x

### PR DESCRIPTION
**Pull request by LaRiffle**
## Description
This PR introduces the following changes
- Add Cuda support for SMPC
- Change encrypted Convolution to use Beaver triples
- Indroduce a new serialization strategy using PyArrow for crypto primitives

#### Current known problems
- BatchNorm uses a trick at conversion (inverses its running var) but this could lead to errors

## Checklist
- [ ] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [ ] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [ ] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [ ] My changes are covered by tests

### CODE REVIEW

File: syft/frameworks/torch/hook/hook.py [hunk 1]
```diff
@@ -714,8 +736,20 @@ def module_encrypt_(nn_self, **kwargs):
             if module_is_missing_grad(nn_self):
                 create_grad_objects(nn_self)
 
-            for element_iter in tensor_iterator(nn_self):
-                for p in element_iter():
+            for element_iter in named_tensor_iterator(nn_self):
+                for name_p, p in element_iter():
+                    if "running_var" in name_p:
```
Comment (gmuraru) [inline syft/frameworks/torch/hook/hook.py:741]: NITPICK: Maybe it would look cleaner to have if like:
```
if "running_var" not in name_p:
    continue

if nn_self.training:
   continue

warnings.warn(...)

````
^^This would be only to don't have so many nested "for, ifs"

File: syft/frameworks/torch/hook/hook.py [hunk 2]
```diff
@@ -808,6 +851,38 @@ def location(nn_self):
...
+                if isinstance(next_node.child, AutogradTensor):
+                    next_node = next_node.child
+                if isinstance(next_node.child, FixedPrecisionTensor):
+                    next_node = next_node.child
+                if isinstance(next_node.child, AdditiveSharingTensor):
+                    raise ValueError(
+                        "Please don't call .eval() on an encrypted model containing "
+                        "a BatchNorm, or fix this."
+                    )
+                # if isinstance(nn_self.running_var.child, (FixedPrecisionTensor, AutogradTensor)):
```
Comment (gmuraru) [inline syft/frameworks/torch/hook/hook.py:881]: Q: This needs to be removed or TODO?

File: syft/frameworks/torch/mpc/primitives.py [hunk 3]
```diff
@@ -204,54 +190,63 @@ def add_primitives(self, types_primitives: dict):
...
-                                    )
-                                current_primitives[i] = tuple(new_prims)
-                            else:
-                                current_primitives[i] = np.concatenate(
-                                    (current_primitives[i], primitive),
-                                    axis=len(primitive.shape) - 1,
-                                )
+                    # This branch never happens with on-the-fly primitives
+                    current_primitives.append(primitives)
+                # print('Added to the store', primitives)
```
Comment (gmuraru) [inline syft/frameworks/torch/mpc/primitives.py:212]: Q: This needs to be removed?

File: syft/serde/compression.py [hunk 4]
```diff
@@ -19,7 +17,8 @@
     LZ4: LZ4.to_bytes(1, byteorder="big"),
     ZLIB: ZLIB.to_bytes(1, byteorder="big"),
 }
-default_compress_scheme = LZ4
+default_compress_scheme = NO_COMPRESSION
```
Comment (gmuraru) [inline syft/serde/compression.py:20]: Q: can't we have only the default one?

File: syft/frameworks/torch/hook/hook.py (hunk 1 above)
Comment (LaRiffle) [reply syft/frameworks/torch/hook/hook.py:741]: hmmm yeah but I need to execute `p.encrypt(inplace=True, **kwargs)`  for all elements of the loop and _after_ the transformation

File: syft/serde/compression.py (hunk 4 above)
Comment (LaRiffle) [reply syft/serde/compression.py:20]: done!

File: syft/frameworks/torch/mpc/primitives.py (hunk 3 above)
Comment (LaRiffle) [reply syft/frameworks/torch/mpc/primitives.py:212]: Done!

File: syft/frameworks/torch/hook/hook.py (hunk 2 above)
Comment (LaRiffle) [reply syft/frameworks/torch/hook/hook.py:881]: Added!

**Review by gmuraru — APPROVED** [review]
